src/Azure/Queues/AzureServiceBusQueue.cs

```diff
@@ -7,73 +7,89 @@
         private readonly int _retries;
+        private readonly RetryPolicy _retryPolicy;
         private readonly TimeSpan _workItemTimeout = TimeSpan.FromMinutes(5);
+        private readonly AsyncLock _lock = new AsyncLock();
 
-        public AzureServiceBusQueue(string connectionString, string queueName = null, int retries = 2, TimeSpan? workItemTimeout = null, bool shouldRecreate = false, RetryPolicy retryPolicy = null, ISerializer serializer = null, IEnumerable<IQueueBehavior<T>> behaviors = null, ILoggerFactory loggerFactory = null) : base(serializer, behaviors, loggerFactory) {
+        public AzureServiceBusQueue(string connectionString, string queueName = null, int retries = 2, TimeSpan? workItemTimeout = null, RetryPolicy retryPolicy = null, ISerializer serializer = null, IEnumerable<IQueueBehavior<T>> behaviors = null, ILoggerFactory loggerFactory = null) : base(serializer, behaviors, loggerFactory) {
+            _connectionString = connectionString;
             _queueName = queueName ?? typeof(T).Name;
             _namespaceManager = NamespaceManager.CreateFromConnectionString(connectionString);
             _retries = retries;
-            if (workItemTimeout.HasValue && workItemTimeout.Value < TimeSpan.FromMinutes(5))
+            _retryPolicy = retryPolicy;
+
+            if (workItemTimeout.HasValue && workItemTimeout.Value < TimeSpan.FromMinutes(5)) {
                 _workItemTimeout = workItemTimeout.Value;
+            }
+        }
 
-            if (_namespaceManager.QueueExists(_queueName) && shouldRecreate)
-                _namespaceManager.DeleteQueue(_queueName);
+        protected override async Task EnsureQueueCreatedAsync(CancellationToken cancellationToken = new CancellationToken()) {
+            if (_queueClient != null) {
+                return;
+            }
 
-            if (!_namespaceManager.QueueExists(_queueName)) {
-                _queueDescription = new QueueDescription(_queueName) {
-                    MaxDeliveryCount = retries + 1,
-                    LockDuration = _workItemTimeout
-                };
-                _namespaceManager.CreateQueue(_queueDescription);
-            } else {
-                _queueDescription = _namespaceManager.GetQueue(_queueName);
+            using (await _lock.LockAsync(cancellationToken)) {
+                if (_queueClient != null) {
+                    return;
+                }
 
-                bool changes = false;
+                QueueDescription queueDescription;
 
-                int newMaxDeliveryCount = retries + 1;
-                if (_queueDescription.MaxDeliveryCount != newMaxDeliveryCount) {
-                    _queueDescription.MaxDeliveryCount = newMaxDeliveryCount;
-                    changes = true;
+                if (!await _namespaceManager.QueueExistsAsync(_queueName).AnyContext()) {
+                    queueDescription = await _namespaceManager.CreateQueueAsync(new QueueDescription(_queueName) {
```

Comment:
You should catch the exception thrown when `CreateQueueAsync` fails because the queue already exists.

We've seen this in our current project using Foundatio: in a multi-process environment, `QueueExistsAsync` can return `false`, for multiple processes, and then they each call `CreateQueueAsync`, but only one will succeed. The rest will fail with either `Microsoft.ServiceBus.Messaging.MessagingException` or `Microsoft.ServiceBus.Messaging.MessagingEntityAlreadyExistsException`.

The correct solution is to catch those exceptions, and then fall back on calling `GetQueueAsync`. If that _also_ fails, then you can bail. :)